### PR TITLE
add end-to-end hello world test

### PR DESCRIPTION
**Pull request by schweitzpgi**

### CODE REVIEW

File: test/Examples/main.c [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+
+/* main entry into PROGRAM */
+void _QQmain();
+
+/* C main stub */
+int main(int argc, char **argv, char **envp)
+{
+  RTNAME(ProgramStart)(argc, argv, envp);
+  _QQmain();
+  RTNAME(ProgramEndStatement)();
```
Comment (jeanPerier) [inline test/Examples/main.c]: I think Peter's design is that the fortran program be called `main` and that the compiler generates the `ProgramStart`/`ProgramEndStatment`. But that also seems fine to start like this for tests.

File: include/flang/Optimizer/Support/InternalNames.h [hunk 2]
```diff
@@ -110,7 +110,8 @@ struct NameUniquer {
                          llvm::StringRef name);
 
   /// Entry point for the PROGRAM (called by the runtime)
-  constexpr static llvm::StringRef doProgramEntry() { return "MAIN_"; }
+  /// Can be overridden with the `--main-entry-name=<name>` option.
```
Comment (jeanPerier) [inline include/flang/Optimizer/Support/InternalNames.h]: +1 for this option

File: lib/Optimizer/Support/InternalNames.cpp [hunk 3]
```diff
@@ -186,6 +189,12 @@ fir::NameUniquer::doVariable(llvm::ArrayRef<llvm::StringRef> modules,
   return result.append(toLower(name));
 }
 
+llvm::StringRef fir::NameUniquer::doProgramEntry() {
+  if (mainEntryName.size())
+    return mainEntryName;
```
Comment (jeanPerier) [inline lib/Optimizer/Support/InternalNames.cpp]: Should this be `_QQ` + mainEntryName so that we retain the info that this is the main program ?

Comment (schweitzpgi) [reply lib/Optimizer/Support/InternalNames.cpp]: My thought is that if someone wants to override the name, then we don't want to get in the way at all.

File: test/Examples/main.c (hunk 1 above)
Comment (schweitzpgi) [reply test/Examples/main.c]: There are trade-offs to consider, but the stub is just to get something limping along for the moment.

**Pull request merged**